cdisutils/parsers.py

```diff
@@ -0,0 +1,80 @@
+import re
+
+try:
+    from urlparse import urlparse
+except ImportError:
+    from urllib.parse import urlparse
+
+
+class S3URLParser:
+    """Adapter for urlparse to facilitate working with S3 urls"""
+    def __init__(self, s3_url):
+        self._s3_url = s3_url
+        self.scheme = None
+        self.netloc = None
+        self.bucket = None
+        self.key = None
+        self._parse()
+
+    def __repr__(self):
+        return "S3URLParser(scheme='{}', netloc='{}', bucket='{}', key='{}')".format(
+            self.scheme,
+            self.netloc,
+            self.bucket,
+            self.key,
+        )
+
+    def _parse(self):
+        """Parsing options for new and old style s3 urls.
+
+        All new buckets created on aws will use the new style of url.
+
+        Old: s3://url/bucket_name/key/name/at/the/end
+        New: https://bucket.url/key/name/at/the/end
+        """
+        parse_object = urlparse(self._s3_url)
+
+        self.scheme = parse_object.scheme
+        self.netloc = parse_object.netloc
+
+        # (scheme, bucket, netloc, path)
+        aws_regex = re.compile(r"^(s3|https?):\/\/([^\.\s]+)\.?(s3.*\.amazonaws\.com)\/(.+)$")
+        aws_match = aws_regex.match(self._s3_url)
+        if aws_match:
+            aws_groups = aws_match.groups()
+
+            self.scheme = aws_groups[0]
+            self.netloc = aws_groups[2]
+
+            if aws_groups[1] != '':
+                # If the bucket is before the netloc, it's a new style url.
+                self.bucket = aws_groups[1]
+                self.key = aws_groups[3]
+            else:
+                # Else get the bucket from the path.
+                path_parts = aws_groups[3].split('/', 1)
+                self.bucket = path_parts[0]
+                self.key = path_parts[1]
+        else:
+            # eg ['', 'bucket', 'key/name/goes/here']
+            path_parts = parse_object.path.split('/', 2)
+
+            # Do not add preceding forward slashes.
+            self.bucket = path_parts[1]
+            self.key = path_parts[2]
```

Comment:
This fails if the URL is totally missing the key (like `s3://ceph/bucket`), but tbh we should raise something in that case, so it's fine.